closed/adds/jdk/src/share/native/jdk/crypto/jniprovider/NativeCrypto.c

```diff
@@ -82,11 +90,30 @@ typedef BIGNUM* OSSL_BN_bin2bn_t (const unsigned char *, int, BIGNUM *);
 typedef void OSSL_BN_set_negative_t (BIGNUM *, int);
 typedef void OSSL_BN_free_t (BIGNUM *);
 
+typedef int OSSL_CRYPTO_num_locks_t();
+typedef void OSSL_CRYPTO_THREADID_set_numeric_t(CRYPTO_THREADID *id, unsigned long val);
+typedef void* OSSL_OPENSSL_malloc_t(size_t num);
+typedef void OSSL_OPENSSL_free_t(void *addr);
```

Comment:
This type is not used, but should be (paired with uses of `OSSL_OPENSSL_malloc`).

Reply:
Please arrange to release resources allocated by `thread_setup()`; i.e. reinstate `OSSL_OPENSSL_free_t` and use it.

Reply:
I still don't see a place where the resources allocated by `thread_setup()` are released.

Reply:
@ashbm5

Reply:
You need to create (and export) a function with this name and signature:
```
void JNICALL JNI_OnUnload(JavaVM * vm, void * reserved)
```
It should perform the necessary cleanup, whether or not `thread_setup()` was called.
It may be called more than once.